sn_networking/src/external_address.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright 2024 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+#![allow(dead_code)]
+
+use crate::{driver::NodeBehaviour, mulitaddr_get_ip};
+use itertools::Itertools;
+use libp2p::{Multiaddr, Swarm};
+use std::{collections::HashMap, net::IpAddr};
+
+const MAX_REPORTS_BEFORE_CONFIRMATION: u8 = 3;
+const MAX_CANDIDATES: usize = 50;
+
+/// Manages the external addresses of a Public node. For a relayed node, the RelayManager should deal with
+/// adding and removing external addresses.
+// TODO:
+// 1. if the max candidate is reached, kick out the oldest candidate sorted by # of reports
+#[derive(Debug)]
+pub struct ExternalAddressManager {
+    /// All the external addresses of the node
+    address_states: Vec<ExternalAddressState>,
```

Comment:
Maybe it's better to use a `HashMap<Multiaddr, State>` here? This would guarantee the uniqueness per element by multiaddr.

Reply:
Yes! Sounds like a good idea!

Reply:
Will make the changes after testing this out!

Reply:
Tried making the change, but I prefer the way with just storing the `Vec<States>`, the state contains the multiaddr and I prefer that one instead of storing it at the key OR duplicating it.